utils.py

```diff
@@ -6,5 +15,109 @@ def filter_country_sort_year(df, country):
     :return: DataFrame
     """
     df = df.loc[df["country_name"] == country]
-    df = df.sort_values(["year"])
+
+    df = df[df.year >= START_YEAR]
+    df = df.sort_values(["year"], ascending=False)
     return df
+
+
+import math
+
+millnames = ["", " K", " M", " B", " T"]
+
+
+def millify(n):
+    n = float(n)
+    millidx = max(
+        0,
+        min(
+            len(millnames) - 1, int(math.floor(0 if n == 0 else math.log10(abs(n)) / 3))
+        ),
+    )
+
+    return "{:.2f}{}".format(n / 10 ** (3 * millidx), millnames[millidx])
+
+
+def handle_empty_plot(fig, text=None):
+    if not text:
+        text = "No Matching Data Found"
+    fig.update_layout(
+        annotations=[
+            dict(
+                text=text,
+                showarrow=False,
+                x=None,
+                y=None,
+            )
+        ],
+    )
+    return fig
+
+
+def get_percentage_change_text(percent):
+    if percent > 0:
+        return f"increased by {percent:.2f}%"
+    return f"decreased by {-1 * percent:.2f}%"
+
+
+def calculate_PCC(df, x_col, y_col):
+    """
+    Calculate the Pearson Correlation Coefficient between two columns
+    :param df: DataFrame
+    :param x_col: str
+    :param y_col: str
+    :return: float
+    """
+    df = df[[x_col, y_col]].dropna()
+    return pearsonr(df[x_col], df[y_col])[0]
+
+
+def get_correlation_text(df, x_col, y_col):
+    """
+    Get the correlation text based on the PCC value
+    :param df: DataFrame
+    :param x_col: str
+    :param y_col: str
+    :return: str
+    """
+    pcc = calculate_PCC(df, x_col, y_col)
+    if pcc > 0:
+        text = "positive and "
+    else:
+        text = "negative and "
+    abs_pcc = abs(pcc)
+
+    for threshold, pcc_text in CORRELATION_THRESHOLDS.items():
+        if abs_pcc < float(threshold):
+            text += pcc_text
+            break
+    return text + "(PCC={:.2f})".format(pcc)
+
+
+def detect_trend(df, x_col):
+    """
+    Detect the trend of the data based on the PCC value
+    :param df: DataFrame
+    :param x_col: str
+    :return: str
+    """
+    pcc = calculate_PCC(df, x_col, "year")
+    abs_pcc = abs(pcc)
+    if abs_pcc > TREND_THRESHOLDS:
+        if pcc > 0:
+            return "an increasing trend"
+        else:
+            return "a decreasing trend"
+
+    return ""
```

Comment:
Maybe we should have a check here for how many data we points we have (either here or in the calculate_PCC function)? If we only had 2 points for computing the PCC then we should probably not say anything about the correlation (since we would always get either +1 or -1)

Reply:
the check is done before calling this function.
https://github.com/weilu/rpf-country-dash/blob/5148a8212bad528e71de7caa4a76efb1c695fe49/pages/education.py#L379-L382